### DISCUSSION TITLE
Can the MFA email extension set a device cookie such that the OTP is not required on every login?

### DISCUSSION THREAD

**Discussion by jfletche** [mesutpiskin/keycloak-2fa-email-authenticator#13] [Q&A]
Hi,

Is there a mechanism such that once a client device has authenticated it can be "remembered" such that the next session login from that same device does not require the OTP?

In other words I authenticate from a device, next time I authenticate on that same device I do not need another OTP as the device is marked as trusted.

Cheers

**Comment by mesutpiskin** [comment]
you can follow to implement this case

Modify your authentication flow: Within your authentication flow, after the user successfully enters the OTP and is authenticated, you can add an action that marks the device as trusted. This can be done by setting a custom attribute or flag on the user's session or user profile, indicating that the device is trusted.

Remember me functionality: Keycloak provides a "Remember Me" feature that allows users to stay authenticated on a device for a longer period without requiring the OTP again. You can enable the "Remember Me" feature in Keycloak's realm settings. This feature typically works by issuing a long-lived browser cookie that stores the user's authentication session. When the user accesses the application again from the same device, the session is restored, and the user is considered authenticated without the need for OTP verification.

Custom device recognition: If you want to implement a custom device recognition mechanism, you can use device fingerprints or other techniques to identify and remember trusted devices. This would require additional development and integration with Keycloak.

**Reply by jfletche** [reply]
Thanks very much for that. Very useful. I will try some of these features out.

Cheers

**Comment by mesutpiskin** [comment]
Keycloak does not have a built-in "remember this device" mechanism for custom authenticators, but you can approximate it using the **Conditional Email OTP** authenticator included in this plugin combined with Keycloak's existing conditional flow support.

The cleanest approach is to use Keycloak's own **"Remember Me"** session combined with a conditional flow:

1. Use the **Conditional Email OTP** authenticator (`email-conditional-authenticator`) instead of the plain Email OTP.
2. Configure the **Skip OTP for Role** or **OTP control User Attribute** options to allow skipping for certain users/conditions.

For true device fingerprinting (cookie-based trust), that would require a custom implementation at the Keycloak level (a separate SPI), as the plugin only handles the OTP step — it has no access to persistent device tokens across sessions.

A simpler alternative: configure the Keycloak realm's **SSO Session Max** (under **Realm Settings → Sessions**) to a long duration. Combined with **Remember Me**, users on trusted devices will stay logged in without re-authenticating (and therefore won't hit the OTP step again until their session expires).
